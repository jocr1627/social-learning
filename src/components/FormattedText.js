import getLinkify from 'linkify-it';
import PropTypes from 'proptypes';
import React, { Component } from 'react';
import tlds from 'tlds';

const linkify = getLinkify();

linkify.tlds(tlds);

let lineBreakKeyIndex = 0;

const insertLineBreaks = (text, elements) => {
  const textBlocks = text.split(/\n|\r/);

  textBlocks.forEach((textBlock, index) => {
    if (textBlock) {
      elements.push(textBlock);
    }

    if (index < textBlocks.length - 1) {
      const brProps = {
        key: lineBreakKeyIndex++,
      };
      const br = <br { ...brProps }/>;

      elements.push(br);
    }
  });
};

const getFormattedElements = (text) => {
  const linkMatches = linkify.match(text) || [];
  let elements = [];
  let currentIndex = 0;

  linkMatches.forEach((match, matchIndex) => {
    const linkProps = {
      href: match.url,
      key: matchIndex,
      target: '_blank',
    };
    const link = <a { ...linkProps }>{ match.url }</a>;
    const rawText = text.slice(currentIndex, match.index);
    
    insertLineBreaks(rawText, elements);

    elements.push(link);
    currentIndex = match.lastIndex;    
  });

  insertLineBreaks(text.slice(currentIndex), elements);

  return elements;
};

export default class FormattedText extends Component {
  static propTypes = {
    text: PropTypes.string.isRequired,
  };

  render() {
    const {
      text,
    } = this.props;
    const formattedElements = getFormattedElements(text);

    return <span>{ formattedElements }</span>;
  }
}
