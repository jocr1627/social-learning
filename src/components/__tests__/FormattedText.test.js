import React from 'react';
import renderer from 'react-test-renderer';

import FormattedText from '../FormattedText';

describe('FormattedText', () => {
  it('replaces newlines with <br/>', () => {
    const text = `
      Consider, once more, the universal cannibalism of the sea; all whose creatures prey upon each other, carrying on eternal war since the world began.
      
      Consider all this; and then turn to the green, gentle, and most docile earth; consider them both, the sea and the land; and do you not find a strange analogy to something in yourself?
    `;
    const formattedTextProps = {
      text,
    };
    const dom = renderer.create(
      <FormattedText { ...formattedTextProps }/>
    ).toJSON();
  
    expect(dom).toMatchSnapshot();
  });

  it('replaces urls with <a/>', () => {
    const text = `
      Try out our recipes at http://delicious-fries.com.
      In need of a change? www.dime-to-exercise.net.
      We knew you'd be back https://www.burgertopia.org/menu?time=lunch
    `;
    const formattedTextProps = {
      text,
    };
    const dom = renderer.create(
      <FormattedText { ...formattedTextProps }/>
    ).toJSON();

    expect(dom).toMatchSnapshot();
  });
});
