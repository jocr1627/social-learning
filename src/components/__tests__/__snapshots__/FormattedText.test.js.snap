// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormattedText replaces newlines with <br/> 1`] = `
<span>
  <br />
        Consider, once more, the universal cannibalism of the sea; all whose creatures prey upon each other, carrying on eternal war since the world began.
  <br />
        
  <br />
        Consider all this; and then turn to the green, gentle, and most docile earth; consider them both, the sea and the land; and do you not find a strange analogy to something in yourself?
  <br />
      
</span>
`;

exports[`FormattedText replaces urls with <a/> 1`] = `
<span>
  <br />
        Try out our recipes at 
  <a
    href="http://delicious-fries.com"
    target="_blank"
  >
    http://delicious-fries.com
  </a>
  .
  <br />
        In need of a change? 
  <a
    href="http://www.dime-to-exercise.net"
    target="_blank"
  >
    http://www.dime-to-exercise.net
  </a>
  .
  <br />
        We knew you'd be back 
  <a
    href="https://www.burgertopia.org/menu?time=lunch"
    target="_blank"
  >
    https://www.burgertopia.org/menu?time=lunch
  </a>
  <br />
      
</span>
`;
